ports/esp32s2/Makefile

```diff
@@ -0,0 +1,255 @@
+# Port-specific
+include mpconfigport.mk
+
+# CircuitPython-specific
+include $(TOP)/py/circuitpy_mpconfig.mk
+
+# qstr definitions (must come before including py.mk)
+QSTR_DEFS = qstrdefsport.h
+
+# include py core make definitions
+include $(TOP)/py/py.mk
+
+include $(TOP)/supervisor/supervisor.mk
+
+# Include make rules and variables common across CircuitPython builds.
+include $(TOP)/py/circuitpy_defns.mk
+
+CROSS_COMPILE = xtensa-esp32s2-elf-
+
+#######################################
+# CFLAGS
+#######################################
+
+INC += -I.
+INC += -I../..
+INC += -I$(BUILD)
+INC += -I$(BUILD)/genhdr
+INC += -I./boards
+INC += -I./boards/$(BOARD)
+INC += -I./peripherals
+INC += -I../../lib/mp-readline
+INC += -I../../lib/tinyusb/src
+INC += -I../../supervisor/shared/usb
+INC += -Iesp-idf/components/freertos/include/freertos
+INC += -Iesp-idf/components/freertos/xtensa/include
+INC += -Iesp-idf/components/esp32s2/include
+INC += -Iesp-idf/components/xtensa/esp32s2/include
+INC += -Iesp-idf/components/esp_common/include
+INC += -Iesp-idf/components/esp_rom/include
+INC += -Iesp-idf/components/xtensa/include
+INC += -Iesp-idf/components/esp_timer/include
+INC += -Iesp-idf/components/soc/include
+INC += -Iesp-idf/components/soc/src/esp32s2/include
+INC += -Iesp-idf/components/soc/soc/esp32s2/include
+INC += -Iesp-idf/components/heap/include
+INC += -Iesp-idf/components/esp_system/include
+INC += -I$(BUILD)/esp-idf/config
+
+CFLAGS += -DHAVE_CONFIG_H \
+          -DMBEDTLS_CONFIG_FILE=\"mbedtls/esp_config.h\" \
+          -DUNITY_INCLUDE_CONFIG_H -DWITH_POSIX
+
+#Debugging/Optimization
+ifeq ($(DEBUG), 1)
+  CFLAGS += -ggdb
+  # You may want to enable these flags to make setting breakpoints easier.
+  # CFLAGS += -fno-inline -fno-ipa-sra
+else
+  CFLAGS += -Os -DNDEBUG -ggdb3
+  # TODO: Test with -flto
+  ### CFLAGS += -flto
+endif
+
+CFLAGS += $(INC) -Werror -Wall -nostdlib $(BASE_CFLAGS) $(C_DEFS) $(CFLAGS_MOD) $(COPT)
+CFLAGS += -DCONFIG_ECC -DCONFIG_IEEE80211W -DCONFIG_SHA256 -DCONFIG_TLSV12 -DCONFIG_WPA3_SAE -DCONFIG_WPS2 -DCONFIG_WPS_PIN -DEAP_MSCHAPv2 -DEAP_PEAP -DEAP_PEER_METHOD -DEAP_TLS -DEAP_TTLS -DESP32_WORKAROUND -DESPRESSIF_USE -DESP_SUPPLICANT -DIEEE8021X_EAPOL -DMBEDTLS_CONFIG_FILE=\"mbedtls/esp_config.h\" -DUSE_WPA2_TASK -DUSE_WPS_TASK -D__ets__
+CFLAGS += -mlongcalls -ffunction-sections -fdata-sections -fstrict-volatile-bitfields -ggdb -Og -std=gnu99 -D_GNU_SOURCE -DIDF_VER=\"v4.2-dev-1128-g6cc269364-dirty\" -DESP_PLATFORM
+
+# LD_FILE := boards/$(BOARD)/$(BOARD)-spi.ld
+
+LDFLAGS = -mlongcalls -nostdlib \
+	-Lesp-idf/components/esp_rom/esp32s2/ld \
+	-L$(BUILD)/esp-idf/esp-idf/esp32s2 \
+	-L$(BUILD)/esp-idf/esp-idf/esp32s2/ld \
+	-Tesp32s2_out.ld \
+	-Tesp32s2.project.ld \
+	-Tesp32s2.rom.ld \
+	-Tesp32s2.rom.libgcc.ld \
+	-Tesp32s2.rom.newlib-data.ld \
+	-Tesp32s2.rom.newlib-funcs.ld \
+	-Tesp32s2.rom.spiflash.ld \
+	-Lesp-idf/components/esp32s2/ld \
+	-Tesp32s2.peripherals.ld \
+	-ucall_user_start_cpu0 \
```

Comment:
answering the question from the livestream: we need to "undefine" functions so that the linker needs to look for them in the static libraries.

If the linker encounters an object file inside a static library, and the object file doesn't resolve any of the undefined references, then the object file is discarded. So we add these -u flags so that linker looks into the archives and includes the object files we want.